design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,499 @@
+apiVersion: apps/v1
+kind: ControllerRevision
+metadata:
+  name: frontend-c8bb659c5 # you could name this anything you wanted, but just appending the semver would be good practice
+revision: 1
+data:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash lscpu"
+```
+
+And it will be recorded in the status of `component.yaml`.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: Component
+metadata:
+  name: frontend
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash lscpu"
+status:
+  latestRevision: frontend-c8bb659c5
+```
+
+
+If you make a change to `component.yaml`:
+
+```shell script
+           cmd:
+-            - "bash lscpu"
++            - "bash top"
+```
+
+A new `ControllerRevision` will be automatically generated:
+
+```shell script
+$ kubectl get ControllerRevisions
+NAMESPACE     NAME                  CONTROLLER               REVISION   AGE
+default       frontend-c8bb659c5    core.oam.dev/component   1          2d15h
+default       frontend-a75588698    core.oam.dev/component   2          2d14h
+
+$ kubectl get ControllerRevisions frontend-a75588698 -o yaml
+apiVersion: apps/v1
+kind: ControllerRevision
+metadata:
+  name: frontend-a75588698 # you could name this anything you wanted, but just appending the semver would be good practice
+revision: 2
+spec:
+  workload:
+    apiVersion: core.oam.dev/v1alpha2
+    kind: ContainerizedWorkload
+    spec:
+      containers:
+      - name: my-cool-workload
+        image: example/very-cool-workload:0.1.2@sha256:verytrustworthyhash
+        cmd:
+        - "bash top"
+```
+
+Now the basic `ControllerRevision` is in place, let's talk about it in Trait and ApplicationConfiguration.
+
+### ControllerRevision could be referenced by Trait 
+
+Some traits can control version of component which means they can specify a component revision and won't be affected
+by the change of component.
+
+For these traits, they have to declare `createRevision` field to be `true` in TraitDefinition like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: fancytraits.core.oam.dev
+spec:
+  createRevision: true
+  definitionRef:
+    name: fancytraits.core.oam.dev
+```
+
+Workload instance's name emitted by OAM runtime will be decided by the `createRevision` field.
+
+- If **NO** traits binding with a component has `createRevision` as true, OAM runtime will let the workload itself to handle
+   versioning things. We will only have one workload instance with the same name of componentName, and update workload directly
+   when component changed.
```

Comment:
what about components that specified with `revisionName `?

Reply:
It will be fixed in that revision, I explained it in latter section